crates/ruff_linter/src/rules/pycodestyle/rules/logical_lines/continued_indentation.rs

```diff
@@ -0,0 +1,804 @@
+                    let diagnostic = Diagnostic::new(
+                        ContinuationUnalignedHanging, // E131
+                        tok.range,
+                    );
+                    context.push_diagnostic(diagnostic);
+                } else {
+                    hangs[bracket_depth] = Some(hang);
+                    // SAFETY: Lines are short, not likely to overflow
+                    if hang > isize::try_from(indent_width.as_usize()).unwrap() {
+                        let diagnostic = Diagnostic::new(
+                            ContinuationOverIndentedHanging, // E126
+                            tok.range,
+                        );
+                        context.push_diagnostic(diagnostic);
+                    } else {
+                        let diagnostic = Diagnostic::new(
+                            ContinuationUnderIndentedHanging, // E121
+                            tok.range,
+                        );
+                        context.push_diagnostic(diagnostic);
+                    }
+                }
+            }
+            prev_newline_token = None;
+        }
+
+        if parens[line_no] > 0
+            && tok.kind() != TokenKind::NonLogicalNewline
+            && tok.kind() != TokenKind::Comment
+            && bracket_indent_level[bracket_depth] == 0
+        {
+            // look for visual indenting
+            bracket_indent_level[bracket_depth] = token_indent;
+            indent_chances.insert(token_indent, VisualIndentType::Normal);
+        } else if matches!(
+            tok.kind(),
+            TokenKind::String | TokenKind::Comment | TokenKind::FStringStart
+        ) {
+            // deal with implicit string concatenation
+            indent_chances.insert(token_indent, VisualIndentType::String);
+        } else if line_no == 0
+            && bracket_depth == 0
+            && matches!(
+                tok.kind(),
+                TokenKind::Assert | TokenKind::Raise | TokenKind::With
+            )
+        {
+            indent_chances.insert(
+                token_indent + tok.range.len().to_usize() + 1,
+                VisualIndentType::Normal,
+            );
+        } else if indent_chances.is_empty()
+            && line_no == 0
+            && bracket_depth == 0
+            && tok.kind() == TokenKind::If
+        {
+            // special case for the 'if' statement because len("if (") == 4
+            indent_chances.insert(
+                token_indent + tok.range.len().to_usize() + 1,
+                VisualIndentType::Normal,
+            );
+        } else if tok.kind() == TokenKind::Colon
+            && logical_line
+                .text_after(tok)
+                .chars()
+                .all(char::is_whitespace)
+        {
+            // If there's only whitespace following this token, it can't be a dictionary construction
+            // TODO: Is this always correct?
+            open_lines[bracket_depth].push(line_no);
+        }
```

Comment:
flagging this to see if there's a better way of doing it (and indeed if it's logically correct)